### PR TITLE
Instrument editor love

### PR DESCRIPTION
**Pull request by charbeljc**
Ok, second batch for the instrument editor, which become finally mostly usable :grin: 

On the technical side, wave widgets now track mouse movements continuously, this allow us to highlight envelope points before we drag / delete them, and see why editing start/loop/end frames in main widget is broken (final PR to come :grinning: )

@theGreatWhiteShark if you could have a look ...
Regards.

### CODE REVIEW

**Comment by charbeljc** [comment]
I have one more commit pending before this pull request is complete ...

**Comment by charbeljc** [comment]
Ok, I'm done with Sample Editing :grinning:

**Comment by theGreatWhiteShark** [comment]
> @theGreatWhiteShark if you could have a look ...
>
> Ok, I'm done with Sample Editing grinning

That's quite fast :smile: . I'll try to take a look within the next days.

**Comment by theGreatWhiteShark** [comment]
Hey @charbeljc ,

As these are improvements rather than bug fixes could you change the target to branch `development`? (Sorry for the inconvenience. This will be straight more straight forward once 1.1 is release and we use `master` for incremental improvements again)

You do not need to close this one and open a new one but can just press the "Edit" button to the right of the title of this PR and select a different branch via a combo box.

**Review by theGreatWhiteShark — COMMENTED** [review]
Thanks for your contribution! I like it a lot!

Could you also have a look at the coding conventions in the  [DEVELOPERS](https://github.com/hydrogen-music/hydrogen/blob/master/DEVELOPERS) file in our repo? I know that our code right now is quite a mess. But if we all stick to the same conventions it will become more easily readable in the future.

I also found two further inconsistencies in the Sample Editor (which are probably not related to your PR)
- open the SampleEditor -> enter a large value in the spinbox of the "Start" and press enter -> both the values in the spinboxes of "Start" and "Loop" get updated but only the "S" line gets moved in the wave display while "L" remains at zero.
- right to the "Play original sample" button there is a the label "new sample length" which seem to show incorrect values until the "E" marker is moved.

File: src/gui/src/SampleEditor/TargetWaveDisplay.cpp [hunk 1]
```diff
@@ -238,117 +240,123 @@ void TargetWaveDisplay::updateDisplay( H2Core::InstrumentLayer *pLayer )
...
+
+		if ((x == 0 || x == UI_WIDTH) && envelope.size() > 2) {
+			// don't remove first and last points if there is more than two points
+			update();
+			return;
+		} else if (envelope.size() == 2) {
+			// if only 2 points, remove them both
+			envelope.clear();
+		} else {
+			envelope.erase( envelope.begin() +  m_nSelectedEnvelopePoint );
```
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/TargetWaveDisplay.cpp:355]: Good catch :grin:, see 29edd84e5db4f6fcaa91ed821d2d84b4d8e915f7

File: src/gui/src/SampleEditor/SampleEditor.h [hunk 2]
```diff
@@ -55,9 +55,10 @@ class SampleEditor : public QDialog, public Ui_SampleEditor_UI, public H2Core::O
 		bool getCloseQuestion();
 		bool returnAllMainWaveDisplayValues();
 		void returnAllTargetDisplayValues();
-		void setTrue();
-		
-		bool m_bSampleEditorStatus;
+		void setUnclean();
+		void setClean();
+
+		bool m_bSampleEditorClean;
```
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/SampleEditor.h]: done, see 397034a258a43d8fa562b62bf388e7dd02a5b0c1

File: src/gui/src/SampleEditor/SampleEditor.cpp [hunk 3]
```diff
@@ -345,8 +344,10 @@ void SampleEditor::on_PrevChangesPushButton_clicked()
 	QApplication::setOverrideCursor(Qt::WaitCursor);
 	getAllLocalFrameInfos();
 	createNewLayer();
-	m_bSampleEditorStatus = true;
+	setClean();
 	QApplication::restoreOverrideCursor();
+	InstrumentEditorPanel *panel = InstrumentEditorPanel::get_instance();
+	panel->selectLayer( panel->getSelectedLayer() ); // reselect layer to trigger update
```
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/SampleEditor.cpp]: see 499fbca0a9972a2176653d61994c9835b1077fd0

File: src/gui/src/SampleEditor/MainSampleWaveDisplay.h [hunk 4]
```diff
@@ -50,17 +58,20 @@ class MainSampleWaveDisplay : public QWidget, public H2Core::Object
 		int		m_nStartFramePosition;
 		int		m_nLoopFramePosition;
 		int		m_nEndFramePosition;
-		bool	m_bMove;
+
 		bool	m_bStartSliderIsMoved;
 		bool	m_bLoopSliderIsMoved;
 		bool	m_bEndSliderIsmoved;
 
+		Slider  m_SelectedSlider;
```
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/MainSampleWaveDisplay.h:66]: done, see 710676298b659cb52a758e3926759870758b629d

File: src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp [hunk 5]
```diff
@@ -283,3 +303,26 @@ void MainSampleWaveDisplay::mouseReleaseEvent(QMouseEvent *ev)
 
 
 
+
+void MainSampleWaveDisplay::chooseSlider(QMouseEvent * ev)
```
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:309]: For the y target coordinates, I choose mostly the y location of the slider's labels (S, L, E), also, the slider selection is done on mouse press and never change while dragging.

Before, the selected slider could change while dragging, furthermore, there was a dead zone between each slider influence.

The previous behavior  could you drive nuts ...

File: src/gui/src/SampleEditor/TargetWaveDisplay.cpp [hunk 6]
```diff
@@ -80,12 +84,31 @@ TargetWaveDisplay::~TargetWaveDisplay()
...
+static void envelope_paint(Sample::VelocityEnvelope &envelope, QPainter &painter,
+	int selected, const QColor & lineColor, const QColor & handleColor, const QColor & selectedColor)
+{
+	if (envelope.empty())
+		return;
+	for ( int i = 0; i < static_cast<int>(envelope.size()) -1; i++){
+		painter.setPen( QPen(lineColor, 1 , Qt::SolidLine) );
+		painter.drawLine( envelope[i]->frame, envelope[i]->value, envelope[i + 1]->frame, envelope[i +1]->value );
+		if ( i == selected )
+			painter.setBrush( selectedColor );
```
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/TargetWaveDisplay.cpp:96]: done, see 2b6ff4cf237317c3c46c68707dc1990318cd0a03

File: src/gui/src/SampleEditor/TargetWaveDisplay.cpp [hunk 7]
```diff
@@ -238,117 +240,123 @@ void TargetWaveDisplay::updateDisplay( H2Core::InstrumentLayer *pLayer )
...
+			} else {
+				envelope.push_back( std::make_unique<EnvelopePoint>( m_nX, m_nY ) );
 			}
+			sort( envelope.begin(), envelope.end(), EnvelopePoint::Comparator() );
+		} else {
+			// move old point to new position
+			update_envelope(envelope, m_nSelectedEnvelopePoint, m_nX, m_nY, m_nSnapRadius);
+		}
+	} else if (ev->button() == Qt::RightButton ) {
+		//remove point
```
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/TargetWaveDisplay.cpp:347]: done also, see 2b6ff4cf237317c3c46c68707dc1990318cd0a03

File: src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp [hunk 8]
```diff
@@ -228,53 +267,34 @@ void MainSampleWaveDisplay::testPosition( QMouseEvent *ev )
 			m_nEndFramePosition = m_nStartFramePosition;
 			m_bEndSliderIsmoved = true;
 		}
-//		update();
 	}
 
 //loopframeposition
-	else if  (ev->y()<=65 ) {
```
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:235]: Yes, this was for consistency with the spinbox behavior. When you try to decrease loop_frame before start_frame, the spinbox clip it at start_frame. Let me know what you think about it.

File: src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp [hunk 9]
```diff
@@ -200,25 +217,47 @@ void MainSampleWaveDisplay::updateDisplay( const QString& filename )
...
 void MainSampleWaveDisplay::mouseMoveEvent(QMouseEvent *ev)
 {
-	testPosition( ev );
+	if (ev->buttons() && Qt::LeftButton) {
+		testPosition( ev );
+	} else {
+		chooseSlider( ev );
+	}
 	update();
+	if (propagate()){
```
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:236]: Hum, you are right ...

[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp]: see cb9866a7ff725a418435133b6480d067d266bae4

**Comment by charbeljc** [comment]
@theGreatWhiteShark Thanks for your review, I addressed most of your comments and I should be really done this time.
Let me know what you think and I will rebase this PR on development branch. Regards,

**Comment by charbeljc** [comment]
> Thanks for your contribution! I like it a lot!
> 
> Could you also have a look at the coding conventions in the [DEVELOPERS](https://github.com/hydrogen-music/hydrogen/blob/master/DEVELOPERS) file in our repo? I know that our code right now is quite a mess. But if we all stick to the same conventions it will become more easily readable in the future.
> 
> I also found two further inconsistencies in the Sample Editor (which are probably not related to your PR)
> 
>     * open the SampleEditor -> enter a large value in the spinbox of the "Start" and press enter -> both the values in the spinboxes of "Start" and "Loop" get updated but only the "S" line gets moved in the wave display while "L" remains at zero.
> 
>     * right to the "Play original sample" button there is a the label "new sample length" which seem to show incorrect values until the "E" marker is moved.

I'll have a look at this later ! (and also, I will try to follow developper's style guide)

File: src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (theGreatWhiteShark) [reply src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:235]: Ah, I see. I think it's fine.

File: src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (theGreatWhiteShark) [reply src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:298]: Definitely. It's a huge improvement. The only thing I found a little bit strange was that clicking in the space between two adjacent sliders in order to move one of them there yields different results depending on the y coordinate of the mouse pointer.

File: src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp [hunk 10]
```diff
@@ -199,26 +218,41 @@ void MainSampleWaveDisplay::updateDisplay( const QString& filename )
...
 void MainSampleWaveDisplay::mouseMoveEvent(QMouseEvent *ev)
 {
-	testPosition( ev );
+	if (ev->buttons() && Qt::LeftButton) {
+		testPosition( ev );
+	} else {
+		chooseSlider( ev );
+	}
 	update();
+	mouseUpdateDone();
```
Comment (theGreatWhiteShark) [inline src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:237]: How come you mark the state of the SampleEditor unclean once the user enters the MainSampleWaveDisplay using the mouse? This makes the "Apply Changes" not flat and indicates something has changed. Changes via drag should be already covered by the call of `mouseUpdateDone` in `mousePressEvent`

**Review by theGreatWhiteShark — COMMENTED** [review]
Nice.

There are still some oddities but they were most probably already present so you do not need to fix them.

- the aforementioned: open the SampleEditor -> enter a large value in the spinbox of the "Start" and press enter -> both the values in the spinboxes of "Start" and "Loop" get updated but only the "S" line gets moved in the wave display while "L" remains at zero.
- Open the SampleEditor -> change something and hit the "Close" button -> a dialog "Unsave changes..." will pop up and the user has to consent to discard the changes -> clicking "Ok" and reopen the SampleEditor by clicking "Edit Layer" again -> the same popup shows up again. I think the second one is a dud.
- When entering a fresh instance of the SampleEditor the value of the `EndFrameSpinBox` looks perfectly fine. But as soon as the mouse pointer enters the `MainSampleWaveDisplay` it is reset to 0 by `SampleEditor::getAllFrameInfos()` what causes some other oddities. The slider, however, stays at it's correct position.

**Comment by theGreatWhiteShark** [comment]
In addition, could you add all the changes in the UX to the [ChangeLog](https://github.com/hydrogen-music/hydrogen/blob/development/ChangeLog) once you rebased on `development`?

File: src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (charbeljc) [reply src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:298]: > Definitely. It's a huge improvement. The only thing I found a little bit strange was that clicking in the space between two adjacent sliders in order to move one of them there yields different results depending on the y coordinate of the mouse pointer.

Yeah, but I think we can get used of this behavior. At least we can make a mental model from it, and having the actual slider selected highlighted helps a lot.

File: src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp (hunk 10 above)
Comment (charbeljc) [reply src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:237]: > How come you mark the state of the SampleEditor unclean once the user enters the MainSampleWaveDisplay using the mouse? This makes the "Apply Changes" not flat and indicates something has changed. Changes via drag should be already covered by the call of `mouseUpdateDone` in `mousePressEvent`

Hum, actually, it does not mark the editor as changed, only setClean()/setUnclean() does that. I should maybe rename `mouseUpdateDone()` to something like `mouseActionDone()`

**Comment by charbeljc** [comment]
> In addition, could you add all the changes in the UX to the [ChangeLog](https://github.com/hydrogen-music/hydrogen/blob/development/ChangeLog) once you rebased on `development`?

Ahem, the ChangeLog... Those are my weak point... :smile:

I just rebased this PR on development branch, see #1302. I would be glad if it could be merged soon, because I have interesting things to come, namely python bindings for the core library, UI as a shared library, with python bindings also, and it starts to make quite a big stack of commits :grin:

**Comment by theGreatWhiteShark** [comment]
> I just rebased this PR on development branch, see #1302. I would be glad if it could be merged soon, because I have interesting things to come,

Sure.

> See #1303 for the bindings grinning

Nice!

**Pull request closed without merging**